### DISCUSSION TITLE
Self-hosted GitLab autorization

### DISCUSSION THREAD

**Discussion by Zerpico** [outline/outline#3217] [Self Hosting] [answered]
_Sorry for my english_.
I'm trying to do an authorization with GitLab.
Installed OutLine via docker with this [manual](https://app.getoutline.com/share/770a97da-13e5-401e-9f8a-37949c19f97e/doc/docker-7pfeLP5a8t).
Create OAuth provider in Gitlab, set values in settings:
```
OIDC_CLIENT_ID=<my_client_id_gitlab>
OIDC_CLIENT_SECRET=<my_secret_gitlab>
OIDC_AUTH_URI=http://gitlab.domen.org/oauth/authorize
OIDC_TOKEN_URI=http://gitlab.domen.org/oauth/token
OIDC_DISPLAY_NAME=GitLab
OIDC_SCOPES="api"
```
On login page after gitlab permission OAuth i get the following message
```Authentication failed – we were unable to sign you in at this time. Please try again.```
And this error in logs:
```
{"error":"Only absolute URLs are supported","stack":"TypeError: Only absolute URLs are supported\n    at getNodeRequestOptions (/opt/outline/node_modules/node-fetch/lib/index.js:1327:9)\n    at /opt/outline/node_modules/node-fetch/lib/index.js:1440:19\n    at new Promise (<anonymous>)\n    at fetch (/opt/outline/node_modules/node-fetch/lib/index.js:1437:9)\n    at fetch (/opt/outline/node_modules/fetch-with-proxy/lib/index.js:123:36)\n    at request (/opt/outline/build/server/utils/passport.js:65:54)\n    at OAuth2Strategy._passportOauth.Strategy.userProfile (/opt/outline/build/server/routes/auth/providers/oidc.js:49:51)\n    at loadIt (/opt/outline/node_modules/passport-oauth2/lib/strategy.js:389:17)\n    at OAuth2Strategy._loadUserProfile (/opt/outline/node_modules/passport-oauth2/lib/strategy.js:404:25)\n    at /opt/outline/node_modules/passport-oauth2/lib/strategy.js:180:16\n    at /opt/outline/node_modules/oauth/lib/oauth2.js:209:7\n    at passBackControl (/opt/outline/node_modules/oauth/lib/oauth2.js:134:9)\n    at IncomingMessage.<anonymous> (/opt/outline/node_modules/oauth/lib/oauth2.js:157:7)\n    at IncomingMessage.emit (node:events:532:35)\n    at IncomingMessage.emit (node:domain:475:12)\n    at endReadableNT (node:internal/streams/readable:1346:12)\n    at processTicksAndRejections (node:internal/process/task_queues:83:21)","level":"error","message":"Error during authentication"}
```
Has anyone set up authorization through Gitlab? Explain what could be wrong

**Comment by tommoor** [comment, marked as answer]
You must also set `OIDC_USERINFO_URI`, I do not know what it should be for GitLab but I do know that some users have succesfully set it up.

**Reply by Zerpico** [reply]
> You must also set `OIDC_USERINFO_URI`, I do not know what it should be for GitLab but I do know that some users have succesfully set it up.

Thank you, i set `OIDC_USERINFO_URI=http://gitlab.domen.org/api/v4/user`.
Now there is no error on the authorization page, but I still cannot log in. The error in the logs is now:

```
{"error":"WHERE parameter \"providerId\" has invalid \"undefined\" value","stack":"Error: WHERE parameter \"providerId\" has invalid \"undefined\" value\n    at PostgresQueryGenerator.whereItemQuery (/opt/outline/node_modules/sequelize/lib/dialects/abstract/query-generator.js:2270:13)\n    at /opt/outline/node_modules/sequelize/lib/dialects/abstract/query-generator.js:2259:25\n    at Array.forEach (<anonymous>)\n    at PostgresQueryGenerator.whereItemsQuery (/opt/outline/node_modules/sequelize/lib/dialects/abstract/query-generator.js:2257:35)\n    at PostgresQueryGenerator.getWhereConditions (/opt/outline/node_modules/sequelize/lib/dialects/abstract/query-generator.js:2675:19)\n    at PostgresQueryGenerator.selectQuery (/opt/outline/node_modules/sequelize/lib/dialects/abstract/query-generator.js:1326:28)\n    at PostgresQueryInterface.select (/opt/outline/node_modules/sequelize/lib/dialects/abstract/query-interface.js:954:27)\n    at Function.findAll (/opt/outline/node_modules/sequelize/lib/model.js:1752:47)\n    at processTicksAndRejections (node:internal/process/task_queues:96:5)\n    at async Function.findOne (/opt/outline/node_modules/sequelize/lib/model.js:1916:12)\n    at async userCreator (/opt/outline/build/server/commands/userCreator.js:27:16)\n    at async accountProvisioner (/opt/outline/build/server/commands/accountProvisioner.js:57:20)\n    at async OAuth2Strategy._verify (/opt/outline/build/server/routes/auth/providers/oidc.js:93:22)","level":"error","message":"Error during authentication"}
```

**Reply by Zerpico** [reply]
In file oidc.ts, try to parse json with the specified variable `OIDC_USERINFO_URI=http://gitlab.org/api/v4/user`.
Attempt to read fields *picture* and *sub* resulted in an error. I change code and it works. Thanks for helping
![image](https://user-images.githubusercontent.com/6038354/157913291-e052e6c4-b971-4f1f-b220-2f1cffd20e67.png)

**Reply by tommoor** [reply]
GitLab is an OIDC provider which means it adheres to the spec – you don't need to fork the code. Try using `https://gitlab.domen.org/oauth/userinfo` as the USERINFO URI

**Reply by Zerpico** [reply]
I tried already. This link doesn't even work.
There are no other links to the profile of an authorized user in json format in gitlab.
I could not use other OpenId, there is no such field "**sub**" anywhere

**Reply by tommoor** [reply]
https://openid.net/specs/openid-connect-basic-1_0-28.html#id_token

Weird. I believe you must have gotten the uri wrong, It's a required part of the spec.

**Reply by deqre** [reply]
this config works for me:

OIDC_AUTH_URI=https://gitlab.example.com/oauth/authorize
OIDC_TOKEN_URI=https://gitlab.example.com/oauth/token
OIDC_USERINFO_URI=https://gitlab.example.com/oauth/userinfo
OIDC_USERNAME_CLAIM=username
OIDC_DISPLAY_NAME=Gitlab
OIDC_SCOPES=openid email

and in gitlab integration scopes: openid email

**Comment by herzkerl** [comment]
Sorry for hijacking—I just tried to implement OIDC authentication via Gitlab with the configuration @deqre posted before, but there’s no authentication button on the login page. I have used Slack before—should Outline (self-hosted) work with two authentication providers?

**Reply by almereyda** [reply]
If you have Slack and OIDC/GitLab configured at the same time in a self-hosted environment, it seems the PR at https://github.com/outline/outline/pull/2986 will help to solve your issue.

**Reply by herzkerl** [reply]
Thank you very much! :)
